src/uv_tcp.c

```diff
@@ -46,13 +40,12 @@ void UvTcpMaybeFireCloseCb(struct UvTcp *t)
         return;
     }
 
-    assert(QUEUE_IS_EMPTY(&t->accepting));
-    assert(QUEUE_IS_EMPTY(&t->connecting));
-
-    if (t->listener.data != NULL) {
+    if (t->listeners) {
         return;
     }
-    if (!QUEUE_IS_EMPTY(&t->aborting)) {
+    assert(QUEUE_IS_EMPTY(&t->accepting));
+
+    if (!QUEUE_IS_EMPTY(&t->connecting) || !QUEUE_IS_EMPTY(&t->aborting)) {
```

Comment:
As said above, the need to check `t->connecting` sounds odd, because it's expected to be empty. Maybe it's a leftover from the previous version with `default_listener`?

Or more probably because the explicit call to `UvTcpMaybeFireCloseCb(t)` in the `else` branch below, where there are no listeners. Since `uvTcpClose` calls `UvTcpListenClose` **before** `UvTcpConnectClose` the `UvTcpMaybeFireCloseCb` callback is called before `UvTcpConnectClose` has been to run, which should not happen.

If that is true, one solution would be to add the `UvTcpMaybeFireCloseCb()` call to then end of `uvTcpClose` instead of `UvTcpListenClose`:

```C
static void uvTcpClose(struct raft_uv_transport *transport,
                       raft_uv_transport_close_cb cb)
{
    struct UvTcp *t = transport->impl;
    assert(!t->closing);
    t->closing = true;
    t->close_cb = cb;
    UvTcpListenClose(t);
    UvTcpConnectClose(t);
    /* Immediately invoke the close callback if there's no
     * asynchronous process to wait for. */
    UvTcpMaybeFireCloseCb(t);
}
```

Reply:
You are totally right I had to replace the assertion by the if due the direct invocation UvTcpMaybeFireCloseCb. As I wrote in a different comment this was required to get all the unit test working. And we always come back to the same problem. Up to now the tcp-listen part of the tcp object **always** contained a tcp handler, which always got closed. As I removed this tcp handle and replaced it by a dynamic created list of tcp handles there is no longer a guaranteed invocation of the UvTcpMaybeFireCloseCb and a lot of the unit tests will fail. This will never happen in any real usage of the raft library as the init will return an error in this case.
I have no problem to move the UvTcpMaybeFireCloseCb, but in this case I still have to replace the assertion by the if as with this movement the UvTcpMaybeFireCloseCb will **always** be invoked synchronously, even if there are listen sockets opened. But at least the execution order for unit test execution and real usage of the library.

Reply:
I think that as long as the `UvTcpMaybeFireCloseCb` call is at the end of `uvTcpClose`, after `UvTcpListenClose` and `UvTcpConnectClose`, then all assertions should work unchanged. Unless I'm missing something.